### ISSUE TITLE
Better documentation on self hosted setup

### ISSUE THREAD

**Issue by cristian-milea** [supabase/supabase#12963]
# Bug report
- kong.yml is created as a directory, although it seems it was resolved about a year ago
- `db` service is written in the docker-compose.yml file and makes it difficult to use the GitHub repository as a starting point for Docker if you want to detach the db, as the tutorial says (ex: in Portainer, with composer override + fork repo is not enough) 
- no clear instructions on how to get the database working with initialization or migrations. 
- not clear what needs to be done in any step, very confusing instructions regarding detached DB
- just trying to start supabase as it is in docker/docker-compose.yml with some env vars doesn't work
- got the studio to work, but gets stuck on default project screen, a lot of 404 errors from `/rest/v1/` and `/api/profile/permissions`
- no explanations on what needs to change in all the env, compose files etc in case your 8000 port is already used on your server. one must search through issues and google and solutions don't work
- very little support on "common issues"
- lots of env variables not explained, have no clue what to change or when or why
- database doesn't connect in the containers, even with the normal database (not detached)

## Describe the bug
I don't think they are bugs, because some people got it working, but the information is so minimal, that's almost impossible to succeed by yourself.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Go to [supabase](https://supabase.com/docs/guides/self-hosting)
2. Try to follow the instructions
3. Battle the Lords of Hell, jump over all the problems, fail, repeat, fail again
4. See all kinds of errors in all supabase created containers
5. Watch your beautiful supabase studio doing nothing, feel sad, write an issue on GitHub 

## Expected behavior
Have a comprehensive step-by-step, maybe some common problems and solutions, explain the detached stuff better etc

## System information

- OS: [e.g. macOS, Windows] Ubuntu 22.04
- Browser (if applies) [e.g. chrome, safari] all
- Version of supabase-js: [e.g. 6.0.2] didn't get there
- Version of Node.js: [e.g. 10.10.0] the one installed from supabase docker-compose

## Additional context
I don't like to say it, but nothing works :(
**Love this project**, _but need more information / documentation. Pretty PLEASE_

**Comment by saltcod** [comment]
ooof, this is a fantastic report. Very much appreciate the time you've taken. I'll see if I can find some help on this one.

**Comment by sweatybridge** [comment]
> - kong.yml is created as a directory, although it seems it was resolved about a year ago

I'm gonna ignore this since it's already resolved.

> - db service is written in the docker-compose.yml file and makes it difficult to use the GitHub repository as a starting point for Docker if you want to detach the db, as the tutorial says (ex: in Portainer, with composer override + fork repo is not enough)
> - no clear instructions on how to get the database working with initialization or migrations.
> - not clear what needs to be done in any step, very confusing instructions regarding detached DB
> - got the studio to work, but gets stuck on default project screen, a lot of 404 errors from /rest/v1/ and /api/profile/permissions

All these are one problem using detached db. You have to run `/docker-entrypoint-initdb.d/migrate.sh` on your detached db to initialise it. You also need to run other scripts in [volumes/db](https://github.com/supabase/supabase/blob/master/docker/docker-compose.yml#L238) to enable those adhoc features.

> - just trying to start supabase as it is in docker/docker-compose.yml with some env vars doesn't work
> - no explanations on what needs to change in all the env, compose files etc in case your 8000 port is already used on your server. one must search through issues and google and solutions don't work
> - lots of env variables not explained, have no clue what to change or when or why
> - database doesn't connect in the containers, even with the normal database (not detached)

Are you saying the default compose config doesn't work for you? Which env did you change? Can you attach logs? Which platform and architecture are you running on?

> - very little support on "common issues"

Generally I'd encourage you to create smaller and more specific issue so the relevant team

**Comment by cristian-milea** [comment]
@saltcod , thank you for trying.
@sweatybridge , no need for any defense here.

1. As the kong.yml was still created as a directory, I don't think you can ignore it. It looks like a regression to me. But it's your choice.
2. This information on how to have a functional detached database is not well documented (although it is recommended to have it detached), and your answer is very vague.
3. If you know how the platform works, it doesn't mean that everybody knows it. This assumption leads to incomplete documentation. Not all of us have the same skills, knowledge or understanding.
4. you completely missed the point of my issue, and that is "incomplete documentation". If your first reaction on reading about a user having problems following Supabase's documentation, to blame the user, I don't want to draw any conclusion here.

Just to reiterate, amazing project and thanks to all of Supabase's team! Excited for the future.

**Comment by cristian-milea** [comment]
An example of a good documentation on .env variables (a point that was ignored in my issue):
https://appwrite.io/docs/environment-variables

**Comment by cristian-milea** [comment]
Another example of a good "how to setup with docker" documentation: https://www.openproject.org/docs/installation-and-operations/installation/docker/ 
It has lots of examples, situations, explanations and step by step.

**Comment by sweatybridge** [comment]
> As the kong.yml was still created as a directory, I don't think you can ignore it. It looks like a regression to me. But it's your choice.

Could you try the following command and paste the output here?

```bash
$ docker exec -it 14bcebde59f2 bash
bash-5.1$ ls -la /var/lib/kong/
total 16
drwxr-xr-x    2 root     root          4096 Mar 14 16:26 .
drwxr-xr-x    1 root     root          4096 Mar 14 16:26 ..
-rw-r--r--    1 kong     nogroup       4222 Jan 10 15:47 kong.yml
```

kong.yml should be mounted as file, at least on amd64 macOS. It would be really helpful if you could provide more information to help reproduce the issue you are experiencing.

**Comment by DerDummePunkt** [comment]
I agree with the general sentiment of @cristian-milea 

These are several topics imho,
I would recommend splitting the issue

#### Split the documentation into a howto for testing/evaluating supabase in docker compose vs running a production stack

To me, the goal or purpose of documenting how to quickly spin up a supabase instance on my local machine (be it via docker-compose or some other means), so that I can start playing with it and get a general understanding of how supabase works, is a different goal from giving me recommendations for how to run supabase in production.

If the goal is to lower the barrier of entry, why make the docs for the former case more complicated then they need to be?
Imho there's no need to talk about decoupling the database when you just want to get users to try out supabase.

#### provide more lightweight way to start a docker compose stack for evaluating a self hosted supabase setup

The main repo alone is currently ~600mb. 
I personally dislike having to clone the complete git repo of a project just to get at some yml files for docker compose and kong and 2-3 sql files for running the initial migrations, just to then have to pull the build artifacts as docker images yet again.

Technically, to get started with supabase in docker-compose, all you'd need are the contens of the `docker` folder, afaik github doesnt offer a convinient way of downloading a folder from a repo

I know a lot of projects do exactly what supabase currently encourages new users to do, even some the @cristian-milea mentioned as better examples, but I disagree with that pratice, it being applied more often doesnt make it better.

If I had to guess, @cristian-milea did you actually clone the repo as state in the docs? Your issue of `kong.yml` being created as a folder is the default behaviour of the docker daemon not being able to find a source file or folder for a volume mount

Again, my guess is @cristian-milea did exactly what I did initially: 
I didnt want to clone the whole repo, because that seemed like overkill, and thought I could just copy and pasted the files from the `docker` folder to get going.
Its very easy to then miss some files however, like the mentioned `kong.yml`, 

My suggestion would be to create a seperate git repo that just contains the files neceesary for getting started with the docker compose stack.
The process of syncing any changes to the included files, like the sql files for getting the postgesql up an running, that occur in the main repo, could be automated via a CI like github actions.

#### better document the configs for the self hosted setup
There just a lot that is missing in the docs.

One example:
* The env example lists parameters for `PHONE_AUTH`, but those params are absent from [the page about the configs for the auth server](https://supabase.com/docs/guides/self-hosting/auth/config)

**Comment by cristian-milea** [comment]
Thanks for clarifying this for me @DerDummePunkt . You are probably right and it makes sense now that you say it, not having kong setup with the volume existing on the host machine could lead to this.

Here are the steps I used to start my self hosted journey of Supabase.
I want to host it on my VPS on Ubuntu 22.04, which currently runs:
Portainer Community Edition 2.17.1
Docker version 23.0.1, build a5ee5b1
docker-compose version 1.29.2, build 5becea4c

Portainer has the ability to run a stack from a GitHub repository. So I forked this repo to my account, and used it to create a docker-compose.override, in order to add env variables for exposing Supabase to my custom domain. Also, I had to modify the original docker-compose to comment out the database, in order to use a detached one.
I created a different stack for supabase/postgres and (somehow) ran the migrations after that (without clear instructions on how to run them).

After this and a lot of modifications to my override file and env variables, I managed to get it working on supabase.mydomain.com but no API calls were working, although Kong was finally setup with a file instead of a directory.

I changed the default port of Kong from 8000 and 8443 to 2000 and 2443, because the other ports are already used on my host. Kong is also exposed with it's own subdomain, on kong-sub.mydomain.com

Based on my setup and the documentation available, setting up the ENV variables was a real pain, because the only instruction on what to change is this:

```
# replace if you intend to use Studio outside of localhost
SUPABASE_PUBLIC_URL=http://localhost:8000 
```

but.. is that port, the port of KONG (after trial and error, I think it is). And if it's kong.. how do you go https?
do I need to change anything else? I don't know nothing about these, should I change them or not:

```
SITE_URL=http://localhost:3000
API_EXTERNAL_URL=http://localhost:8000
```


To sum this up. This was not a straight forward process. It was a lot of trial and error, and this is mainly because I wanted to go "detached", as recommended, thinking that I want to have it on my domain and consume Supabase from my apps.

**Comment by alexkreidler** [comment]
I'd also recommend that we make a list somewhere of the differences between Self-Hosted and Cloud, because currently [auth settings](https://github.com/orgs/supabase/discussions/4444) (which was pretty critical for me trying to set up a 3rd party OAuth integration) and [edge functions](https://github.com/supabase/supabase/issues/10319) are not supported on self-hosted, and up until last year, [storage](https://github.com/supabase/supabase/pull/8623) was not either.

**Comment by Hi-ImKyle** [comment]
Without sounding like an ass it's almost as if the idea was to make it intentionally half assed to get you to give into buying the cloud version which I get that you guys deserve to be paid but at that point why wouldn't I just to Firebase. Only reason why I thought about Supabase was because of the option to self host it but I can't do that without jumping through several hoops that are on fire and I'm doused in gasoline.

**Comment by Hi-ImKyle** [comment]
Would be nice if the self hosted docs got a touch up. I think I'm on my 3rd or 4th try at trying to get it all to work fine without issues. Without sounding like an ass, it's almost like they were intentionally made sparse and crappy to get you to buy the cloud version.

**Comment by rienheuver** [comment]
I find myself in the same situation currently. Really trying to get the self-hosted working but running into so many problems... have been trying to get the detached database working for a couple days, but have given up for now. Currently trying to get email working, but I'm just not receiving emails (and don't get any errors either). Improved documentation would help a lot. If anyone succeeded in this mystical journey, perhaps write a blog detailing how you got through it? Would love to do so myself, but simply haven't slain the beast yet

**Comment by davidofug** [comment]
I followed this blog (https://www.linode.com/docs/guides/installing-supabase/) to setup self hosted supabase on GCP but am struggling to point it to the domain.
Indeed the documentation is like 5% of what has to be done.
An improvement would really help if it's the intention of the Supabase's business model.

**Comment by adarshaketh** [comment]
> Would be nice if the self hosted docs got a touch up. I think I'm on my 3rd or 4th try at trying to get it all to work fine without issues. Without sounding like an ass, it's almost like they were intentionally made sparse and crappy to get you to buy the cloud version.

true

**Comment by nokia6290** [comment]
> https://appwrite.io/docs/environment-variables

Thank you.

After struglling with non-existant and sort of low quality documentation, I will give a go with AppWrite open source project.
First impression - they have a documentation which is already better than supabase...
